### ISSUE TITLE
Implement git as a state store

### ISSUE THREAD

**Issue by kris-nova** [kris-nova/kubicorn#135]
We would like to use git as a state store.
This probably means using regular old git, as well as calling APIs like GitHub

[An State Store is defined as the interface with the following functions...](https://github.com/kris-nova/kubicorn/blob/master/state/interface.go).
State Stores are located in the [`state`](https://github.com/kris-nova/kubicorn/tree/master/state) package.

Currently, we have two implementations that could serve you as an example:
* [YAML implementation](https://github.com/kris-nova/kubicorn/tree/master/state/fs),
* [JSON implementation](https://github.com/kris-nova/kubicorn/tree/master/state/jsonfs).

Also, you'll have to update the [`cmd`](https://github.com/kris-nova/kubicorn/tree/master/cmd) pacakge with new state store option (e.g. [this is one of the places where it should be added](https://github.com/kris-nova/kubicorn/blob/master/cmd/create.go#L170)).

If you'd like to learn more about `kubicorn` project, you can read [The Project Walkthrough](http://kubicorn.io/documentation/readme.html).
If you have any question, comment on the issue, join us on the `#kubicorn` channel in the [Gophers Slack community](https://invite.slack.golangbridge.org/) or open a PR. We'll always do everything we can to get the PR though.

Similar issue for S3 buckets -- #417.
Edit: [Implementation idea](https://github.com/kris-nova/kubicorn/issues/135#issuecomment-337587022).

**Comment by mariomazo** [comment]
@kris-nova does it make sense to also add s3?

**Comment by danicat** [comment]
@kris-nova the link to the project walkthrough in the issue description is broken (small typo on the `project` word) http://kubicorn.io/documentation/projetct-walkthrough.html -> http://kubicorn.io/documentation/project-walkthrough.html

**Comment by xmudrii** [comment]
@danicat thank you very much for reporting this. I'll go ahead and update all the links.

**Comment by Ch3ck** [comment]
@xmudrii @kris-nova I'll like to work on this

**Comment by xmudrii** [comment]
@Ch3ck sounds good to me. Do you need some help?

**Comment by Ch3ck** [comment]
Not yet, I'll look through the notes from @kris-nova and let you know if I need anything else

**Comment by xmudrii** [comment]
/cc @Ch3ck @kris-nova 

I'll try to sketch out implementation for Git state store in this comment. Here's the idea...

We want to store states in an Git repo. For now, let's focus just on pure Git, so no GitHub APIs and similar stuff, just pure git. Here's an sample repo -- https://github.com/kris-nova/example-state
As you see, it has multiple states in one repo.

So, let's look for [FS JSON implementation](https://github.com/kris-nova/kubicorn/tree/master/state/jsonfs) for example. 

What would be different (sorted by functions):
* `NewJSONFileSystemStore`
  * This one could stay the same.
* `Exists`
  * For now, let's stay same, we'll assume it exists in on the remote.
* `write`
  * If directory doesn't exist, create it and initialize git repo as well (`git init`).
* `Read`
  * It could stay the same.
* `ReadStore`
  * Same..
* `Commit`
  * We should `git commit` here and probably `git push`.
* `BytesToCluster`
  * Same.
* `List`
  * Same.

Tl;dr: for now let's change: `write`, `Read`, `Commit`. Others are good to go. Basically, Git state store could be implemented same as classic one, just with added initialization and pushing.

To make it easier for you now, I'll not sketch out GitHub State Store implementation. I still has to see what their API offers and how we can utilize it. For now, get this done and once we merged it, we'll continue with other state stores.

**Comment by Ch3ck** [comment]
Perfect @xmudrii this clarifies everything I had in mind.
